src/python/pants/java/jar/jar_dependency.py

```diff
@@ -130,18 +164,14 @@ def transitive(self):
   def copy(self, **replacements):
```

Comment:
This is why it's sometimes useful to use things like `DatatypeMixin`, because we might not have had to make any of the changes to this file at all except for moving `__new__` to `__init__` if we had continued to use the `DatatypeMixin` interface with `.copy()` and `._asdict()`, which abstracts over the specific way data is represented. Would you be open to it if I submitted a followup PR that demonstrates what I'm talking about?

Reply:
With the below code, we only had to change line 167 from `self._asdict()` to  `dataclasses.asdict(self)`.

All the other changes were simply refactors in realizing that `excludes` will already be processed through the `__init__` so there was no need to explicitly call `JarDependency._prepare_excludes`.

--

With `DatatypeMixin` exposing `.copy()` and `._asdict()`, these same behaviors are implemented through `dataclass.asdict()` and `dataclass.replace()`. Preferably we will only have one and only one way to implement these two behaviors. `DatatypeMixin` and those two functions have been useful and were a great utility for the codebase. Now that we have an stdlib version (only possible as of Py3.6), I advocate for going with the solution of leveraging the std lib for better integration with tooling, less code for us to maintain, and lower barrier to entry for new contributors.